### ISSUE TITLE
Add store accounts

### ISSUE THREAD

**Issue by Aninstance** [Aninstance/simple-stock-management#2]
#Preliminary scope

## Backend
Add database for each 'store manager' user. Change stock xfer process to add units to store database, rather than simply delete from warehouse database (as currently the case). Add a 'stocktake' process to calculate a report, as detailed below.

### Database schema
Duplication of warehouse database schema, with exception of:
```
+ 'opening_stock' column
+ 'closing_stock' column
+ 'xfer_price' column (unit price of stock as xferred)
+ 'selling_price' column (current unit price of stock as sold in the shop)
- 'unit_price' column
```

#### Recorded data
- Opening stock (opening_stock as of datetime of last stocktake)
- Closing stock (closing_stock as of datetime 'now' at current stocktake process)

##### Values for each line 
a) Differential for *all* sales & loss (opening_stock - closing_stock)
b) Xfer price value for all units lost or sold ((opening_stock - closing_stock) * xfer_price)
c) Selling price value for all units lost or sold ((opening_stock - closing_stock) * selling_price)
d) Differential between total xfer price & total selling price, for a line ( (c) - (b) ) *See note (1)

##### Values for total store
e) Total xfer price value of all stock lines sold or lost since last stocktake process (Σ(b))
f) Total selling price value of all stock lines sold or lost since last stocktake process (Σ(c))
g) Total differential for all stock lines, between xfer price & selling price (Σ(d)) *See note (1)

### Database maintenance processes
- Following completion of stocktake process, closing_stock data is copied to opening_stock, then emptied.

## UI
Build a dedicated UI for store databases, to allow store managers to:
- View their own store's stock holding record 
- Submit a stock count (closing_stock) & trigger the generation of stocktake reports

## Notes
1. Whilst the differential may be indicative of a store's sales performance, it cannot be considered "profit", because actual sales, shrinkage and loss (both known and unknown) are not recorded. Therefore, the differential is not synonymous with sales and profit in this model. It may merely suggest that, had the differential consisted entirely of sales at the selling price - with no units being subject to loss or shrinkage - the differential would have been representative of "store profit" under those improbable circumstances.

**Comment by Aninstance** [comment]
#Preliminary scope

##Backend

Add database for each 'store manager' user. Change stock xfer process to add units to store database, rather than simply delete from warehouse database (as currently the case). Add a 'stocktake' process to calculate a report, as detailed below.

###Database schema
- 'opening_stock' column
- 'working_stock' column (current stock levels, as displayed in UI and open to +- adjustment by store managers, to record sales, shrinkage, etc)
- 'working_sales' column (total of units of a line recorded sold since last stocktake process)
- 'working_write_off' column (total of units of a line recorded as written off due to damage, loss or theft) 
- 'closing_stock' column
- 'xfer_price' column (unit price of stock as xferred)
- 'selling_price' column (current unit price of stock as sold in the shop)

###Report generation

####Recorded data
- Opening stock (opening_stock as of datetime of last stocktake)
- Closing stock (closing_stock as of datetime 'now' at current stocktake process)
- Units expected sold (working_sales).

####Processed data

#####Values for each line 
a) Number of *unknown* sales & loss (working_stock ∆ closing_stock)
b) Number of *all* sales & loss (opening_stock ∆ closing_stock)
c) Xfer value of all units recorded as sold since last stocktake process (xfer_price * working_sales)
d) Xfer value of all units recorded as write-offs since last stocktake process (xfer_price * working_write_off)
e) Selling total value of all units recorded as sold since last stocktake process (selling_price * working_sales)

#####Values for total store
f) Total xfer value of all stock sold or lost since last stocktake process { Σ(b) }
g) Total xfer value of all stock recorded as sold since last stocktake process { Σ (c) }
h) Total selling value of all stock recorded as sold since last stocktake process { Σ(e) }
i) Total xfer value of all stock recorded as write-offs since last stocktake process { Σ(d) }
j) Total "profit" { (g) ∆  (h) }  *see note (1).
k) Total value differential between all xferred stock and recorded sales + recorded write-offs [may include both unrecorded write-offs and unrecorded sales] { (f) ∆ ((g) + (i)) }

###Database maintenance processes
- Following completion of stocktake process, closing_stock data is copied to opening_stock & working_stock, with closing_stock, working_sales and working_write_off columns being emptied

## UI

- Build a dedicated UI for store databases, to allow store managers to record:
  - Unit sales
  - Manual adjustment (shrinkage, unexpected 'recovered stock', etc)
  - Generation of stocktake reports

##Notes
1. The accuracy of profit calculation would be impacted by changes to the selling price during the period between running stocktake processes. Suppose a shop xfered {n} units at {i} xfer_price, sold {a} units at {y} then dropped the price and sold the remaining {n-a} units at {p}. The profit {T} could then be calculated thus:  T=((a*y)+((n-a)*p))-(n*i). However, the system would have no way of knowing how many were sold before or after the price change. Solutions include:
  a) Trigger generation of stocktake report each time a price is changed on a product (thus resetting the opening stock)
  b) Record sale price for each sale as it occurs (rather than simply be inferred through stock levels). This option would require evolving the app towards a POS system, which may constitute scope-creep at this time.

**Comment by Aninstance** [comment]
- Accounts database holds unique record rows for combination of: user, xfer price & total units. This allows for the same item (SKU) to be held for a single user multiple times if the xfer price was different, therefore permitting value calculations to account for fluctuations in xfer price.

**Comment by Aninstance** [comment]
Released in version 4.0.0
